README
NAME
    Text::Levenshtein::Damerau - Damerau Levenshtein edit distance.

SYNOPSIS
            use Text::Levenshtein::Damerau;
            use warnings;
            use strict;

            my @targets = ('fuor','xr','fourrrr','fo');

            # Initialize Text::Levenshtein::Damerau object with text to compare against
            my $tld = Text::Levenshtein::Damerau->new('four');

            print $tld->dld($targets[0]);
            # prints 1

            my $tld = $tld->dld({ list => \@targets });
            print $tld->{'fuor'};
            # prints 1

            print $tld->dld_best_match({ list => \@targets });
            # prints fuor

            print $tld->dld_best_distance({ list => \@targets });
            # prints 1


            # or even more simply
            use Text::Levenshtein::Damerau qw/edistance/;
            use warnings;
            use strict;
        
            print edistance('Neil','Niel');
            # prints 1

DESCRIPTION
    Returns the true Damerau Levenshtein edit distance of strings with
    adjacent transpositions. Useful for fuzzy matching, DNA variation
    metrics, and fraud detection.

    Defaults to using Pure Perl Text::Levenshtein::Damerau::PP, but has an
    XS addon Text::Levenshtein::Damerau::XS for massive speed imrovements.
    Works correctly with utf8 if backend supports it; known to work with
    "Text::Levenshtein::Damerau::PP" and "Text::Levenshtein::Damerau::XS".

            use utf8;
            my $tld = Text::Levenshtein::Damerau->new('ⓕⓞⓤⓡ');
            print $tld->dld('ⓕⓤⓞⓡ');
            # prints 1

CONSTRUCTOR
  new
    Creates and returns a "Text::Levenshtein::Damerau" object. Takes a
    scalar with the text (source) you want to compare against.

            my $tld = Text::Levenshtein::Damerau->new('Neil');
            # Creates a new Text::Levenshtein::Damerau object $tld

METHODS
  $tld->dld
    Scalar Argument: Takes a string to compare with.

    Returns: an integer representing the edit distance between the source
    and the passed argument.

    Hashref Argument: Takes a hashref containing:

    *   list => \@array (array ref of strings to compare with)

    *   *OPTIONAL* max_distance => $int (only return results with $int
        distance or less).

    *   *OPTIONAL* backend => 'Some::Module::its_function' Any module that
        will take 2 arguments and returns an int. If the module fails to
        load, the function doesn't exist, or the function doesn't return a
        number when passed 2 strings, then "backend" remains unchanged.

                # Override defaults and use Text::Levenshtein::Damerau::PP's pp_edistance()
                $tld->dld({ list=> \@list, backend => 'Text::Levenshtein::Damerau::PP::pp_edistance');

                # Override defaults and use Text::Levenshtein::Damerau::XS's xs_edistance()
                use Text::Levenshtein::Damerau;
                requires Text::Levenshtein::Damerau::XS;
                ...
                $tld->dld({ list=> \@list, backend => 'Text::Levenshtein::Damerau::XS::xs_edistance');

    Returns: hashref with each word from the passed list as keys, and their
    edit distance (if less than max_distance, which is unlimited by
    default).

            my $tld = Text::Levenshtein::Damerau->new('Neil');
            print $tld->dld( 'Niel' );
            # prints 1

            #or if you want to check the distance of various items in a list

            my @names_list = ('Niel','Jack');
            my $tld = Text::Levenshtein::Damerau->new('Neil');
            my $d_ref = $tld->dld({ list=> \@names_list }); # pass a list, returns a hash ref
            print $d_ref->{'Niel'}; #prints 1
            print $d_ref->{'Jack'}; #prints 4

  $tld->dld_best_match
    Argument: an array reference of strings.

    Returns: the string with the smallest edit distance between the source
    and the array of strings passed.

    Takes distance of $tld source against every item in @targets, then
    returns the string of the best match.

            my $tld = Text::Levenshtein::Damerau->new('Neil');
            my @name_spellings = ('Niel','Neell','KNiel');
            print $tld->dld_best_match({ list=> \@name_spellings });
            # prints Niel

  $tld->dld_best_distance
    Arguments: an array reference of strings.

    Returns: the smallest edit distance between the source and the array
    reference of strings passed.

    Takes distance of $tld source against every item in the passed array,
    then returns the smallest edit distance.

            my $tld = Text::Levenshtein::Damerau->new('Neil');
            my @name_spellings = ('Niel','Neell','KNiel');
            print $tld->dld_best_distance({ list => \@name_spellings });
            # prints 1

EXPORTABLE METHODS
  edistance
    Arguments: source string and target string.

    *   *OPTIONAL 3rd argument* int (max distance; only return results with
        $int distance or less). 0 = unlimited. Default = 0.

    Returns: int that represents the edit distance between the two argument.
    -1 if max distance is set and reached.

    Wrapper function to take the edit distance between a source and target
    string. It will attempt to use, in order:

    *   Text::Levenshtein::Damerau::XS xs_edistance

    *   Text::Levenshtein::Damerau::PP pp_edistance

            use Text::Levenshtein::Damerau qw/edistance/;
            print edistance('Neil','Niel');
            # prints 1

SEE ALSO
    *   <https://github.com/ugexe/Text--Levenshtein--Damerau> *repository*

    *   <http://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance>
        *damerau levenshtein explaination*

    *   Text::Fuzzy *regular levenshtein distance*

BUGS
    Please report bugs to:

    <https://rt.cpan.org/Public/Dist/Display.html?Name=Text-Levenshtein-Dame
    rau>

AUTHOR
    Nick Logan <ug@skunkds.com>

LICENSE AND COPYRIGHT
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module Text::Levenshtein::Damerau

0.39 ?
	- Exporter 'import' 5.6 hack removed
	- List::Util min check removed (used full time)

0.38 Wed Jan 16 14:52:44 2013
	- Variable name clarifications
	- ::PP refactoring/optimizations
	- Declaration changes

0.37 Thu Jan 10 00:18:41 2013
	- Optimizations in ::PP

0.36 Thu Dec 13 00:45:41
	- MIN_PERL_VERSION added back to Makefile using different number format

0.35 Sat Dec 08 23:40:10 2012
	- Support for Perls earlier than 5.8 dropped (ha!)

0.34 Thu Dec 06 02:18:37 2012
	- another utf8 update for 5.6

0.33 Sun Nov 30 10:53:13 2012
	- utf8 changes for a Perl 5.6 test failure

0.32 Sun Nov 25 12:31:42 2012
	- Small optimization in backend selection
	- Exporter import error in older Exporter/Perl fixed

0.31 Sun Nov 18 18:14:31 2012
	- Moved pod tests to /xt
	- Doc updates

0.30 Wed Nov 13 17:34:19 2012
	- max distance added to stop calculations at a certain distance

0.29 Sun Oct 28 02:08:19 2012
	- Hardcoded Makefile.PL values

0.28 Sat Oct 27 20:33:14 2012
	- Document updates

0.27 Sat Oct 27 20:33:14 2012
	- Modified _set_backend to test passed module::function for valid return value, ignoring the change request on failure.

0.26 Sat Oct 27 18:29:57 2012
	- Fixed pod utf8 stuff

0.25 Sat Oct 27 15:58:22 2012
	- Document updates

0.24 Sat Oct 27 02:40:38 2012
	- Removed List::Util from ::PP
	- Fixed dld_best_match so it correctly passed the entire argument hash ref
	- Allow changing of backend at run time via dld
	- Document updates

0.23 Fri Oct 26 17:39:51 2012
	- Moved edistance to ::PP sub module. Adding a ::XS distro engine to do the heavy lifting of edistance in C if available.
	- Lots of refactoring 

0.19 Thu Oct 25 00:02:39 2012
	- Test consolidation; All tests use Test::More, removed Test::Base tests
	- Passes use strict by checking dld if its a reference
	- Perl::Tidy'd
	- Pure perl functionality finalized, working on Inline::C backend to do heavy lifting

0.18 Tues Oct 23 23:47:03 2012
	- MORE TESTS
	- fixed error with dld_best_distance
	- I'm sure CPAN authors everywhere are laughing at the number of releases per day :)

0.17 Tues Oct 23 21:23:17 2012
	- Documentation clarifications

0.16 Tues Oct 23 19:23:30 2012
	- Documentation updates
	- Argument/return changes

0.15 Tues Oct 23 13:20:42 2012
	- Documentation updates

0.14 Tues Oct 23 01:51:29 2012
	- Fixed documentation formatting ruining source
	- Sorry CPAN! :(

0.12 Mon Oct 22 23:28:40 2012
	- Documentation updates
	- Returned exportable edistance() function

0.11 Mon Oct 22 19:15:12 2012
	- Completely recoded to be more OO
	- Added ability to retrieve best match from a list
	- Added github repository to metadata
	- Added: utf8 tests, CaSE tests, pod_coverage tests
	- Added List::Util back

0.09 Sun Oct 21 19:52:16 2012
	- Fixed file creation time error with cpan installation

0.08 Sun Oct 21 18:54:48 2012
	- Documentation fix
	- Removed List::Util in favor of local min subroutine
	- return 0 without doing calculations if strings match

0.07 Sun Oct 21 14:35:11 2012
	- use utf8;
	- Added more tests
	- Makefile updated: require List::Util, test_requires Test::More

0.06 Sat Oct 20 16:40:19 2012
	- Added function dld (mirror of edistance) for clarity
	- Modified documentation
	- Added tests using Test::Base and Test::More
	- Switched to Module::Install   

0.05 Sat Sep 20 19:32:28 2012
	- Corrected directory structure to install with cpanm


0.04 Sat Jun 20 14:34:43 2012
	- original version; created by ExtUtils::ModuleMaker





Makefile.PL
use 5.008_008; # for utf8, sorry legacy Perls
use strict;
use warnings; 
use ExtUtils::MakeMaker;

my $pm =   'lib/Text/Levenshtein/Damerau.pm';
my $repo = 'https://github.com/ugexe/Text--Levenshtein--Damerau';
my $bugs = 'https://rt.cpan.org/Public/Dist/Display.html?Name=Text-Levenshtein-Damerau';
 
WriteMakefile (  
    NAME => 'Text::Levenshtein::Damerau',
    VERSION_FROM => $pm, 
    ABSTRACT_FROM => $pm,
    AUTHOR => 'Nick Logan <ugexe@cpan.org>',
    LICENSE => 'perl',
    MIN_PERL_VERSION => '5.008008',
    PREREQ_PM => {
	'List::Util'		=> 0,
    },
    BUILD_REQUIRES => {
	'Test::More'		=> 0,
    },
    META_MERGE => {
        resources => {
            repository => $repo, 
            bugtracker => "$bugs",
        },
        provides => {
            'Text::Levenshtein::Damerau' => {
                file    => 'lib/Text/Levenshtein/Damerau.pm',
                version => '0.40',
            },
            'Text::Levenshtein::Damerau::PP' => {
                file    => 'lib/Text/Levenshtein/Damerau/PP.pm',
                version => '0.24',
            },
        },

    },

    depend => { '$(FIRST_MAKEFILE)' => '$(VERSION_FROM)' },
);


lib/Text/Levenshtein/Damerau/PP.pm
package Text::Levenshtein::Damerau::PP;
use 5.008_008;    # for utf8, sorry legacy Perls
use strict;
use List::Util qw/min/;
require Exporter;
 
our @ISA = qw(Exporter);
our @EXPORT_OK = qw/pp_edistance/;
our $VERSION   = '0.24';

sub pp_edistance {
    # Does the actual calculation on a pair of strings
    my ( $source, $target, $max_distance ) = @_;
    $max_distance = int($max_distance || 0);

    my $source_length = length($source) || 0;
    my $target_length = length($target) || 0;
    return ($source_length?$source_length:$target_length) if(!$target_length || !$source_length);

    my $lengths_max = $source_length + $target_length;
    my $dictionary_count;    #create dictionary to keep character count
    my $swap_count;          
    my $swap_score;          
    my $target_char_count;   
    my $source_index;
    my $target_index;
    my @scores;              

    # init values outside of work loops
    $scores[0][0] = $scores[1][0] = $scores[0][1] = $lengths_max;
    $scores[1][1] = 0;

    # Work Loops
    foreach $source_index ( 1 .. $source_length ) {
        $swap_count = 0;
        $dictionary_count->{ substr( $source, $source_index - 1, 1 ) } = 0;
        $scores[ $source_index + 1 ][1] = $source_index;
        $scores[ $source_index + 1 ][0] = $lengths_max;

        foreach $target_index ( 1 .. $target_length ) {
            if ( $source_index == 1 ) {
                $dictionary_count->{ substr( $target, $target_index - 1, 1 ) } = 0;
                $scores[1][ $target_index + 1 ] = $target_index;
                $scores[0][ $target_index + 1 ] = $lengths_max;
            }

            $target_char_count =
              $dictionary_count->{ substr( $target, $target_index - 1, 1 ) };
	     $swap_score = $scores[$target_char_count][$swap_count] +
                  ( $source_index - $target_char_count - 1 ) + 1 +
                  ( $target_index - $swap_count - 1 );

            if (
                substr( $source, $source_index - 1, 1 ) ne
                substr( $target, $target_index - 1, 1 ) )
            {
                $scores[ $source_index + 1 ][ $target_index + 1 ] = min(
                    $scores[$source_index][$target_index]+1,
                    $scores[ $source_index + 1 ][$target_index]+1,
                    $scores[$source_index][ $target_index + 1 ]+1,
                    $swap_score
                );
            }
            else {
                $swap_count = $target_index;

                $scores[ $source_index + 1 ][ $target_index + 1 ] = min(
                  $scores[$source_index][$target_index], $swap_score
                );
            }
        }

        #unless ( $max_distance == 0 || $max_distance >= $scores[ $source_index + 1 ][ $target_length + 1 ] )
        #{
        #    return -1;
        #}

        $dictionary_count->{ substr( $source, $source_index - 1, 1 ) } =
          $source_index;
    }
 
    my $score = $scores[ $source_length + 1 ][ $target_length + 1 ];
    return ($max_distance != 0 && $max_distance < $score)?-1:$score;
}
 
1;

__END__

=encoding utf8

=head1 NAME

Text::Levenshtein::Damerau::PP - Pure Perl Damerau Levenshtein edit distance.

=head1 SYNOPSIS

	use Text::Levenshtein::Damerau::PP qw/pp_edistance/;

	print pp_edistance('Neil','Niel');
	# prints 1

=head1 DESCRIPTION

Returns the true Damerau Levenshtein edit distance of strings with adjacent transpositions. Pure Perl implementation. Works correctly with utf8.

	use Text::Levenshtein::Damerau::PP qw/pp_edistance/;
	use utf8;

	pp_edistance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ'), 
	# prints 1

=head1 METHODS

=head2 pp_edistance

Arguments: source string and target string. 

=over

=item * I<OPTIONAL 3rd argument> int (max distance; only return results with $int distance or less). 0 = unlimited. Default = 0.

=back

Returns: int that represents the edit distance between the two argument. Stops calculations and returns -1 if max distance is set and reached.

Function to take the edit distance between a source and target string. Contains the actual algorithm implementation. 

	use Text::Levenshtein::Damerau::PP qw/pp_edistance/;
	print pp_edistance('Neil','Niel');
	# prints 1

	print pp_edistance('Neil','Nielx',1);
	# prints -1


=head1 SEE ALSO

=over 4

=item * L<Text::Levenshtein::Damerau>

=item * L<Text::Levenshtein::Damerau::XS>

=back

=head1 BUGS

Please report bugs to:

L<https://rt.cpan.org/Public/Dist/Display.html?Name=Text-Levenshtein-Damerau>

=head1 AUTHOR

Nick Logan <F<ug@skunkds.com>>

=head1 LICENSE AND COPYRIGHT

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut




lib/Text/Levenshtein/Damerau.pm
package Text::Levenshtein::Damerau;
use 5.008_008; # for utf8, sorry legacy Perls
use strict;
use utf8;
use List::Util qw/reduce/;
require Exporter;

our @ISA = qw(Exporter); 
our @EXPORT_OK = qw/edistance/;
our $VERSION   = '0.40';
 
# To XS or not to XS...
unless ( _set_backend('Text::Levenshtein::Damerau::XS::xs_edistance') ) {
    _set_backend('Text::Levenshtein::Damerau::PP::pp_edistance');
}

sub new {
    my $class = shift;
    my $self  = {};

    $self->{'source'}       = shift;
    $self->{'max_distance'} = 0;
    bless( $self, $class );

    return $self;
}

sub dld {
    my $self = shift;
    my $args = shift;

    if ( !ref $args ) {
        return edistance( $self->{'source'}, $args, );
    }
    elsif ( ref $args->{'list'} eq ref [] ) {
        my $target_score;

        if ( defined( $args->{'backend'} ) ) {
            _set_backend( $args->{'backend'} );
        }

        foreach my $target ( @{ $args->{'list'} } ) {
            my $ed =
              edistance( $self->{'source'}, $target, $args->{max_distance} );
            $target_score->{$target} = $ed if ( $ed >= 0 );
        }

        return $target_score;

    }
}
 
sub dld_best_match {
    my $self = shift;
    my $args = shift;

    if ( defined( $args->{'list'} ) ) {
        my $hash_ref = $self->dld($args);

        #Get the hashref key with the smallest value
        return reduce { $hash_ref->{$a} < $hash_ref->{$b} ? $a : $b }
        keys %{$hash_ref};
    } 
}

sub dld_best_distance {
    my $self = shift;
    my $args = shift;
 
    if ( defined( $args->{'list'} ) ) {
        return $self->dld( $self->dld_best_match($args) );
    }
}

sub _set_backend {
    my $be = my $mod = shift;
    $mod =~ s/^(.*)::.*?$/$1/;

    local $@;    
    eval "require $mod";
    unless ($@) {
	 eval "defined &$be";
        unless($@) {
            # We welcome our new edistance overlord
            *edistance = \&$be;
	     return 1;
        }
    }
    
    return 0;
}

1;

__END__

=encoding utf8

=head1 NAME

Text::Levenshtein::Damerau - Damerau Levenshtein edit distance.

=head1 SYNOPSIS

	use Text::Levenshtein::Damerau;
	use warnings;
	use strict;

	my @targets = ('fuor','xr','fourrrr','fo');

	# Initialize Text::Levenshtein::Damerau object with text to compare against
	my $tld = Text::Levenshtein::Damerau->new('four');

	print $tld->dld($targets[0]);
	# prints 1

	my $tld = $tld->dld({ list => \@targets });
	print $tld->{'fuor'};
	# prints 1

	print $tld->dld_best_match({ list => \@targets });
	# prints fuor

	print $tld->dld_best_distance({ list => \@targets });
 	# prints 1


	# or even more simply
	use Text::Levenshtein::Damerau qw/edistance/;
	use warnings;
	use strict;
	
	print edistance('Neil','Niel');
	# prints 1

=head1 DESCRIPTION

Returns the true Damerau Levenshtein edit distance of strings with adjacent transpositions. Useful for fuzzy matching, DNA variation metrics, and fraud detection.

Defaults to using Pure Perl L<Text::Levenshtein::Damerau::PP>, but has an XS addon L<Text::Levenshtein::Damerau::XS> for massive speed imrovements. Works correctly with utf8 if backend supports it; known to work with C<Text::Levenshtein::Damerau::PP> and C<Text::Levenshtein::Damerau::XS>.

	use utf8;
	my $tld = Text::Levenshtein::Damerau->new('ⓕⓞⓤⓡ');
	print $tld->dld('ⓕⓤⓞⓡ');
	# prints 1

=head1 CONSTRUCTOR

=head2 new

Creates and returns a C<Text::Levenshtein::Damerau> object. Takes a scalar with the text (source) you want to compare against. 

	my $tld = Text::Levenshtein::Damerau->new('Neil');
	# Creates a new Text::Levenshtein::Damerau object $tld

=head1 METHODS

=head2 $tld->dld

B<Scalar> Argument: Takes a string to compare with.

Returns: an integer representing the edit distance between the source and the passed argument.

B<Hashref> Argument: Takes a hashref containing:

=over 4

=item * list => \@array (array ref of strings to compare with)

=item * I<OPTIONAL> max_distance => $int (only return results with $int distance or less).

=item * I<OPTIONAL> backend => 'Some::Module::its_function' Any module that will take 2 arguments and returns an int. If the module fails to load, the function doesn't exist, or the function doesn't return a number when passed 2 strings, then C<backend> remains unchanged. 

	# Override defaults and use Text::Levenshtein::Damerau::PP's pp_edistance()
	$tld->dld({ list=> \@list, backend => 'Text::Levenshtein::Damerau::PP::pp_edistance');

	# Override defaults and use Text::Levenshtein::Damerau::XS's xs_edistance()
	use Text::Levenshtein::Damerau;
	requires Text::Levenshtein::Damerau::XS;
	...
	$tld->dld({ list=> \@list, backend => 'Text::Levenshtein::Damerau::XS::xs_edistance');

=back

Returns: hashref with each word from the passed list as keys, and their edit distance (if less than max_distance, which is unlimited by default).

	my $tld = Text::Levenshtein::Damerau->new('Neil');
	print $tld->dld( 'Niel' );
	# prints 1

	#or if you want to check the distance of various items in a list

	my @names_list = ('Niel','Jack');
	my $tld = Text::Levenshtein::Damerau->new('Neil');
	my $d_ref = $tld->dld({ list=> \@names_list }); # pass a list, returns a hash ref
	print $d_ref->{'Niel'}; #prints 1
	print $d_ref->{'Jack'}; #prints 4

=head2 $tld->dld_best_match

Argument: an array reference of strings.

Returns: the string with the smallest edit distance between the source and the array of strings passed.

Takes distance of $tld source against every item in @targets, then returns the string of the best match.

	my $tld = Text::Levenshtein::Damerau->new('Neil');
	my @name_spellings = ('Niel','Neell','KNiel');
	print $tld->dld_best_match({ list=> \@name_spellings });
	# prints Niel

=head2 $tld->dld_best_distance

Arguments: an array reference of strings.

Returns: the smallest edit distance between the source and the array reference of strings passed.

Takes distance of $tld source against every item in the passed array, then returns the smallest edit distance.

	my $tld = Text::Levenshtein::Damerau->new('Neil');
	my @name_spellings = ('Niel','Neell','KNiel');
	print $tld->dld_best_distance({ list => \@name_spellings });
	# prints 1

=head1 EXPORTABLE METHODS

=head2 edistance

Arguments: source string and target string.

=over

=item * I<OPTIONAL 3rd argument> int (max distance; only return results with $int distance or less). 0 = unlimited. Default = 0.

=back

Returns: int that represents the edit distance between the two argument. -1 if max distance is set and reached.

Wrapper function to take the edit distance between a source and target string. It will attempt to use, in order: 

=over 4

=item * L<Text::Levenshtein::Damerau::XS> B<xs_edistance>

=item * L<Text::Levenshtein::Damerau::PP> B<pp_edistance>

=back

	use Text::Levenshtein::Damerau qw/edistance/;
	print edistance('Neil','Niel');
	# prints 1

=head1 SEE ALSO

=over 4

=item * L<https://github.com/ugexe/Text--Levenshtein--Damerau> I<Repository>

=item * L<http://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance> I<Damerau levenshtein explanation>

=item * L<Text::Fuzzy> I<Regular levenshtein distance>

=back

=head1 BUGS

Please report bugs to:

L<https://rt.cpan.org/Public/Dist/Display.html?Name=Text-Levenshtein-Damerau>

=head1 AUTHOR

Nick Logan <F<ug@skunkds.com>>

=head1 LICENSE AND COPYRIGHT

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut


